File: src/main/leetcode/set100/set160/p162/Solution.java
package set100.set160.p162;

/**
 * Created by wangsenyuan on 8/27/16.
 */
public class Solution {
    public static void main(String[] args) {
        int[] nums = {1, 2};
        System.out.println(new Solution().findPeakElement(nums));
    }

    public int findPeakElement(int[] num) {
        int l = 0;
        int r = num.length - 1;
        while (l < r) {
            int mid = (l + r) / 2;

            if (num[mid] < num[mid + 1]) {
                l = mid + 1;
            } else {
                r = mid;
            }
        }
        return l;
    }
}

File: src/main/leetcode/set100/set160/p162/Solution.scala
package set100.set160.p162

object Solution {
  def findPeakElement(nums: Array[Int]): Int = {
    val n = nums.length
    if (n == 1) {
      nums(0)
    } else {
      def go(left: Int, right: Int): Int = {
        if (left == right) {
          left
        } else {
          val mid = (left + right) / 2
          if (nums(mid) < nums(mid + 1)) {
            go(mid + 1, right)
          } else {
            go(left, mid)
          }
        }
      }

      go(0, n - 1)
    }
  }
}
